### DISCUSSION TITLE
Is there a list of trace context propagators that the Java client supports?

### DISCUSSION THREAD

**Discussion by joe-elliott** [open-telemetry/opentelemetry-java-instrumentation#3860] [Q&A] [answered]
I'm looking for documentation that lists the various ways that the Java client can propagate trace context (HTTP, GRPC, ??).

In particular I'm wondering if it automatically propagates through Kafka.

**Comment by mateuszrzeszutek** [comment, marked as answer]
We don't have that documented because we (usually) implement context propagation in all libraries that we instrument - you can assume that all HTTP/PRC/messaging libs listed [here](https://github.com/open-telemetry/opentelemetry-java-instrumentation/blob/main/docs/supported-libraries.md#libraries--frameworks) propagate context properly.

Kafka, however, is a know source of context propagation problems 😅  Because of its poll-based consumer API it's pretty hard to correctly wrap a CONSUMER span around message processing, and depending on the way you're using it you might or might not get your context propagated correctly. If you're able to, please try the agent out - and if you find any context propagation issues with Kafka, don't hesitate to create a bug and attach an example of how you use the Kafka consumer.

**Reply by joe-elliott** [reply]
So context is propagated through Kafka, but depending on the way it's being used it's possible that it won't propagate? Is there any more details about the right or wrong way to use it? or should i just give it a shot?

**Reply by mateuszrzeszutek** [reply]
`KafkaConsumer#poll()` returns a `ConsumerRecords`, which is essentially an `Iterable` (or it returns an `Iterable` with its `records()` method). As long as you iterate over it "normally" it'll work - we start the span on `next()` and end it on `hasNext()`, so if you remember to call `hasNext()` at the end (e.g. `hasNext()`->`next()`->`hasNext()`->...->`hasNext()`) all the spans will get started and ended properly.

**Reply by mateuszrzeszutek** [reply]
If you make a copy of the list, use `.get()` or something other than iteration you'll run into problems with that though.

**Reply by joe-elliott** [reply]
Thank you for the help!
